<a href="https://colab.research.google.com/github/Yamito548/Act-3.3/blob/main/Actividad_3_3_%E2%80%93_IRIS_con_NaiveBayes_y_Cross_Validation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Ejercicicio 3.3** by Dylan J. Ramírez Bermúdez
Dataset pingüinos

# Cross validation

In [ ]:
#Datasets
import seaborn as sns
#maninuplacion de datos
import pandas as pd
import numpy as np
#Gráficos
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from matplotlib.colors import ListedColormap
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import accuracy_score
#Clasificadores
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB, ComplementNB
import matplotlib.pyplot as plt
#Dataset pingüinos
penguin = sns.load_dataset("penguins")
penguin

In [ ]:
#Rellenamos los datos faltantes.
penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']] = \
    penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']].fillna(penguin.mean())

<ipython-input-11-43fa0ba2d099>:3: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']].fillna(penguin.mean())


# Cambiar valores a 1

In [ ]:
le = LabelEncoder()
penguin['sex'] = le.fit_transform(penguin['sex'])

In [ ]:
penguin['sex'] = penguin['sex'].fillna(penguin['sex'].mode().iloc[0])

# MultinomialNB

# Truco cross validation, cortito

In [ ]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = MultinomialNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))
#En estos casos se divide 5 veces, y se entrena. Luego se hace media de todos los resultados.

Resultados de Cross-Validation:
[0.79710145 0.76811594 0.8115942  0.8115942  0.86764706]
Precisión media: 0.81


# GaussianNB

In [ ]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = GaussianNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.97101449 0.95652174 0.95652174 0.98550725 0.95588235]
Precisión media: 0.97


# BernoulliNB

In [ ]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = BernoulliNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.44927536 0.44927536 0.43478261 0.43478261 0.44117647]
Precisión media: 0.44


# ComplementNB

In [ ]:
# Definir las características (X) y la variable objetivo (y)
X = penguin[['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']]
y = penguin['species']

# Crear el clasificador MultinomialNB()
clf = ComplementNB()

# Realizar cross-validation con 5 folds
cv_scores = cross_val_score(clf, X, y, cv=5)

# Imprimir los resultados
print("Resultados de Cross-Validation:")
print(cv_scores)
print("Precisión media: {:.2f}".format(cv_scores.mean()))

Resultados de Cross-Validation:
[0.69565217 0.71014493 0.69565217 0.66666667 0.73529412]
Precisión media: 0.70


# Cross validation a mano

In [ ]:
#Dividamos los datos de entrenamiento
X_train, X_test, y_train, y_test = \
        train_test_split(X, y, test_size=.3, random_state=42)

# Clasificadores
names = ["GaussianNB", "MultiNomialNB", 'BernouilliNB', 'ComplementNB']
classifiers = [GaussianNB(), MultinomialNB(), BernoulliNB(), ComplementNB()]

for name, clf in zip(names, classifiers):

  # Entrenadr el modelo
  clf.fit(X_train, y_train)
  # PREDIT AND SCORE
  score = clf.score(X_test, y_test) #Returns the mean accuracy on the given test data and labels.
  print ("Modelo: %s = %6.2f" % (name, score))

Modelo: GaussianNB =   0.96
Modelo: MultiNomialNB =   0.82
Modelo: BernouilliNB =   0.48
Modelo: ComplementNB =   0.67


En este caso, se puede ver que al usar dos formas diferentes de usar cross validation obtenemos resultados un poco diferentes. Por ejemplo que en el que es más sencillo me dió en complementNB un score de 0.70 mientras que el que es más a mano me dio un 0.67. Siendo un poco mejor, y eso pasa con todos. No creo que esto signifique el 100% es mejor el 1º que el 2º habría que revisar bien como hace los calculos que usa el de cross_val_score.